models/classes/runner/OfflineQtiRunnerService.php

```diff
@@ -0,0 +1,147 @@
+<?php
+/**
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; under version 2
+ * of the License (non-upgradable).
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2019 (original work) Open Assessment Technologies SA ;
+ */
+/**
+ * @author Péter Halász <peter@taotesting.com>
+ */
+
+namespace oat\taoQtiTest\models\runner;
+
+use oat\oatbox\service\ConfigurableService;
+use oat\taoQtiTest\models\runner\map\RunnerMap;
+
+/**
+ * Service class for the offline version of Qti Test Runner
+ */
+class OfflineQtiRunnerService extends ConfigurableService
+{
+    const SERVICE_ID = 'taoQtiTest/OfflineQtiRunnerService';
+
+    /**
+     * Returns an array of items, containing also confident data, like branching and response processing rules
+     *
+     * @param RunnerServiceContext $serviceContext
+     * @return array
+     * @throws \common_Exception
+     * @throws \common_exception_Error
+     * @throws \common_exception_InconsistentData
+     * @throws \common_exception_InvalidArgumentType
+     */
+    public function getItems($serviceContext)
+    {
+        $this->getRunnerService()->assertQtiRunnerServiceContext($serviceContext);
```

Comment:
There are a lot of ServiceLocator calls. I know that it's legacy stuff, but maybe it would be more performant to cache it as a class property. Opinion?

Reply:
I've checked and the services get cached in the ServiceManager and the `get()` call will load them from the config only for the first time.